### ISSUE TITLE
CA2000: Not firing when using interfaces

### ISSUE THREAD

**Issue by CodeBlanch** [dotnet/roslyn-analyzers#5330]
### Analyzer

**Diagnostic ID**: [CA2000](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2000): `Dispose objects before losing scope`

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: 5.0.3 (Latest)

### Describe the bug

The [IMemoryCache.CreateEntry](https://github.com/dotnet/runtime/blob/09c146c815d8a21f36cb138a2487c1b50877a0b0/src/libraries/Microsoft.Extensions.Caching.Abstractions/src/IMemoryCache.cs#L26) API returns an [ICacheEntry](https://github.com/dotnet/runtime/blob/38ae818d8768a8fb5009349955eb572096da5f80/src/libraries/Microsoft.Extensions.Caching.Abstractions/src/ICacheEntry.cs#L13) which implements `IDisposable`. I noticed developers not disposing of the returned instances and CA2000 not firing to remind them.

### Steps To Reproduce

Here's a test that reproduces the issue. When using `MemoryCache` directly CA2000 fires. When using the `IMemoryCache` interface, it doesn't.

```csharp
	using System;

	interface ICacheEntry : IDisposable
	{
	}

	class CacheEntry : ICacheEntry
	{
		public void Dispose()
		{
		}
	}

	interface IMemoryCache
	{
		ICacheEntry CreateEntry();
	}

	class MemoryCache : IMemoryCache
	{
		public ICacheEntry CreateEntry()
		{
			return new CacheEntry();
		}
	}

	class Test
	{
		void M1()
		{
			MemoryCache memoryCache = new MemoryCache();
			memoryCache.CreateEntry(); // CA2000 warning generated here

			IMemoryCache interface_memoryCache = new MemoryCache();
			interface_memoryCache.CreateEntry(); // No warning generated
		}
	}
```

### Expected behavior

CA2000 warning.

### Actual behavior

No warning.

**Comment by ranma42** [comment]
IIUC CA2000 currently considers:
 - explicit creation via constructor
 - creation via two special cases https://github.com/dotnet/roslyn-analyzers/blob/086d1cc888ca0d8c96b94444935f488fcaf3fe98/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/DisposeAnalysis/DisposeAnalysis.DisposeDataFlowOperationVisitor.cs#L259-L262

Would it be ok to consider as "disposable creation" all methods (both static and instance) that return an `IsDisposable` object?
This would possibly increase the false positives, but I guess returning a disposable object is commonly associated to transferring its ownership (for example in factories), so it would also possibly decrease the false negatives.

Alternatively, I see that there have been proposals for making ownership more explicit via attributes, such as https://github.com/dotnet/roslyn-analyzers/issues/5138 and https://github.com/dotnet/runtime/issues/29631 but I see that there has not been a conclusive decision yet (maybe they are still being discussed/designed internally?)

TL;DR: I'd love to provide the wanted help, but I am unsure whether the approach I have in mind is adequate.

**Comment by ranma42** [comment]
fun fact: because of the interaction between the special case and extension methods, some (instance-based) patterns are already detected:

```csharp
	using System;

	interface ICacheEntry : IDisposable
	{
	}

	class CacheEntry : ICacheEntry
	{
		public void Dispose()
		{
		}
	}

	interface IMemoryCache
	{
		ICacheEntry CreateEntry();
	}

	class MemoryCache : IMemoryCache
	{
		public ICacheEntry CreateEntry()
		{
			return new CacheEntry();
		}
	}

	static class MemoryExtension
	{
		public static ICacheEntry Create(this IMemoryCache cache)
		{
			return cache.CreateEntry();
		}
	}

	class Test
	{
		void M1()
		{
			MemoryCache memoryCache = new MemoryCache();
			memoryCache.CreateEntry(); // CA2000 warning generated here
			memoryCache.Create(); // CA2000 warning generated here

			IMemoryCache interface_memoryCache = new MemoryCache();
			interface_memoryCache.CreateEntry(); // No warning generated
			interface_memoryCache.Create(); // CA2000 warning generated here
		}
	}
```

**Comment by CodeBlanch** [comment]
@ranma42 

>IIUC CA2000 currently considers:
>
>* explicit creation via constructor
>* creation via two special cases

I don't think that totally explains this (or at least I'm not understanding):

```csharp
	MemoryCache memoryCache = new MemoryCache();
	memoryCache.CreateEntry(); // CA2000 warning generated here

	IMemoryCache interface_memoryCache = new MemoryCache();
	interface_memoryCache.CreateEntry(); // No warning generated
```

The only difference between the two (that I can tell) is in the first snippet `CreateEntry` is invoked on a concrete type (`MemoryCache `) and in the second `CreateEntry` is invoked on an interface (`IMemoryCache`).

**Comment by ranma42** [comment]
@CodeBlanch IIUC the difference you observe comes from the `new` in `MemoryCache.CreateEntry` (the analyzer is doing an interprocedural analysis)

**Comment by CodeBlanch** [comment]
@ranma42 Got it. So you are saying it will actually look at the `MemoryCache.CreateEntry` implementation and sees the `new` in the body.

In that case...

> Would it be ok to consider as "disposable creation" all methods (both static and instance) that return an IsDisposable object?
This would possibly increase the false positives, but I guess returning a disposable object is commonly associated to transferring its ownership (for example in factories), so it would also possibly decrease the false negatives.

I'm far from an expert in how these analyzers work but IMO the above seems like a more simple and robust solution. If a method returns `IDisposable` I would take that as indication that the caller now owns it (and therefore the responsibility of cleaning it up).

**Comment by colgreen** [comment]
I've just encountered this. I have defined a type that implements IDisposable, and was hoping that CA2000 would detected when I create instances and forget to dispose them, but alas there is no analyser warning/suggestion.

Also note that CA2000 is disabled by default in Visual Studio 2022, so I enabled it as a suggestion, like so in `.editorconfig`:

    dotnet_diagnostic.CA2000.severity = suggestion  # Dispose objects before losing scope.

Somewhat related, but CA2213 (Disposable fields should be disposed) is also disabled by default.

**Comment by jerviscui** [comment]
same question!
The problem is well reproduced, as long as the IDispose is returned by an interface method, it will not trigger the CA2000.

```csharp
private readonly IReturnDisposable _returnInterface;

public async Task Return_Ingerface_Async_Test()
{
    // interface 不会提示 CA2000
    var a = await _returnInterface.ReturnDisposableAsync();
}

public void Return_Ingerface_Sync_Test()
{
    // interface 不会提示 CA2000
    var a = _returnInterface.ReturnDisposable();
}

public interface IReturnDisposable
{
    public BaseClassWithFinalizer ReturnDisposable();

    public Task<BaseClassWithFinalizer> ReturnDisposableAsync();
}

public class BaseClassWithFinalizer : IDisposable, IAsyncDisposable
{}
```

**Comment by jerviscui** [comment]
> fun fact: because of the interaction between the special case and extension methods, some (instance-based) patterns are already detected:
> 
> 	using System;
> 
> 	interface ICacheEntry : IDisposable
> 	{
> 	}
> 
> 	class CacheEntry : ICacheEntry
> 	{
> 		public void Dispose()
> 		{
> 		}
> 	}
> 
> 	interface IMemoryCache
> 	{
> 		ICacheEntry CreateEntry();
> 	}
> 
> 	class MemoryCache : IMemoryCache
> 	{
> 		public ICacheEntry CreateEntry()
> 		{
> 			return new CacheEntry();
> 		}
> 	}
> 
> 	static class MemoryExtension
> 	{
> 		public static ICacheEntry Create(this IMemoryCache cache)
> 		{
> 			return cache.CreateEntry();
> 		}
> 	}
> 
> 	class Test
> 	{
> 		void M1()
> 		{
> 			MemoryCache memoryCache = new MemoryCache();
> 			memoryCache.CreateEntry(); // CA2000 warning generated here
> 			memoryCache.Create(); // CA2000 warning generated here
> 
> 			IMemoryCache interface_memoryCache = new MemoryCache();
> 			interface_memoryCache.CreateEntry(); // No warning generated
> 			interface_memoryCache.Create(); // CA2000 warning generated here
> 		}
> 	}

The result of my testing is that the extension method wrapper also fails to trigger the CA2000.
@ranma42

**Comment by jerviscui** [comment]
Asynchronous methods returning `Task<IDisposable>` cannot generate CA2000
